pymongoexplain/explainable_collection.py

```diff
@@ -0,0 +1,118 @@
+# Copyright 2020-present MongoDB, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import pymongo
+from bson.son import SON
+
+class BaseCommand():
+    def __init__(self, dictionary):
```

Comment:
I suggest we change this API to also accept the command name here. That we get_SON can return a SON() where the command name appears first. This way we only need to use SON in one place.

Reply:
Because of the peculiarities of each command's structure, I don't think it makes sense to store the command name because there will have to be custom SON objects created for many of the classes anyways. For example,
```
return_dictionary = SON([("update", collection.name), ("updates", [SON([("q", SON(filter)), ("u", SON(update))])])])
```
Versus
```
self.dictionary = SON([("aggregate", collection.name), ("pipeline", pipeline)])
```
We would have to have this class specific code anyways, so I personally think it makes more sense to simply insert the command name there rather than have it be an attribute of the base class.

Reply:
You don't need to wrap every dictionary with SON, only the command itself needs to be ordered. My idea is that if you pass the command name to BaseCommand you can use standard `dict` everywhere else and use SON in one place instead of sprinkling SON all around the codebase. Like this:
```
class BaseCommand():
    def __init__(self, command_name, dictionary):
        self.command_name = command_name
        self.dictionary = dictionary
    def get_SON(self):
        # Ensure the command name is the first key, the order of the other
        # command arguments does not matter. 
        cmd = SON([(self.command_name, 1)])
        cmd.update(self.dictionary)
        return cmd
```

Reply:
That makes sense. I have committed changes to address this.